Proposal: Luanti generative AI policy (#17352)
minetest/minetest: doc/ai_policy.md

The most important requirement when using AI tools in Luanti is to understand
and take responsibility for your _entire_ contribution, regardless of whether
AI tooling was used, and to be respectful to others in your usage of AI tools.

That said, the following specific limitations are put in place:

- AI may not be used to generate substantial pieces of code or documentation.
  AI assistance may be used for simple tasks, such as code completion,
  find-and-replace tasks, or generating boilerplate code. In particular,
  "vibe-coding" or autonomous AI usage is completely unacceptable.
- Do not use AI-generated text when communicating with other humans. Nobody
  wants to make a comment and get a machine-generated response: this is basic
  respect. That said, using AI tools to assist with grammar or translations on
  human-written text is perfectly fine.
- AI may not be used to generate art, music, sounds, or any other media. Many
  people feel very strongly about how machines should be used to participate in
  the creative and artistic processes, and we as the Luanti maintainers respect
  those views highly.

Since the state of AI tooling is constantly changing, we may lift some

> restrictions on AI-generated code in the future.

Reviewer: Not just lift but also reinforce (includes adding edge-cases which we have yet not thought of).

```suggestion
Since the state of AI tooling is constantly changing, these restrictions
may be lifted or reinforced in the future.
```

## Permissible uses of AI

In general, any other usage of AI is permissible. For example, here are some
perfectly licit uses of generative AI:

- Exploring the Luanti codebase in a fuzzy way or understanding code
  architecture quickly with AI tooling.
- Using an LLM as a powerful search engine or knowledge base for coding

>   purposes, keeping in mind that you must take responsibility for truly

Reviewer: ```suggestion
  purposes. Keep in mind that you must take responsibility for truly
```
? I am not a native English speaker, hence this suggestion might be incorrect.

Reviewer 2: I prefer the separate sentence as well.

Author: I simplified this one further to shorten the sentence.

  understanding your contribution.
- Having an LLM review your code or check for potential bugs in order to
  improve the quality of your contribution. Note that all PRs must still be

>   fully reviewed and approved by a human maintainer.

Reviewer 3: Would it make sense to add a link to https://docs.luanti.org/for-engine-devs/git-guidelines/? That would make it clear that this is a general requirement and not specific to AI.

Author: In light of https://github.com/luanti-org/luanti/pull/17352#discussion_r3605561533, maybe I should remove this sentence altogether. The git guidelines are already plenty clear about the requirement for core devs to review all PRs.

Reviewer 4: Shouldn't this mention review by the contributor rather than maintainer?

Author: I was referring to core devs in this sentence, but naturally the contributor has to stand by their contribution as well.

Author: I removed the sentence entirely since it's already redundant with the git guidelines. It's only relevant to core devs anyway, so doesn't need to be in the AI policy.

- Using an LLM to debug an issue or search for security vulnerabilities, but
  bear in mind that AIs may not be used for human-to-human communication! Make

>   sure to understand the AI's analysis completely and communicate it yourself.

Reviewer: ```suggestion
- Using an LLM to debug an issue or search for security vulnerabilities, but make
  sure to understand the AI's analysis completely and communicate it yourself.
```

IMO reminders can be nice to have but they do make the text longer. I do not like long texts.

Author: I was wondering if the text was too long. I'll try to trim it down and reduce some of the redundancy.

Author: Shortened.

>   sure to understand the AI's analysis completely and communicate it yourself.

Reviewer 4: If they want to report a security vulnerability, or any bug for that matter, they must not only *understand* it, but also confirm that it actually exists and is a bug.

Author: Changed the word from "understand" to "verify".
